Eclipse 2018 12 compatibility

k-oliver opened the pull request:
Version 1.1.0 to be used in Eclipse 2018-12 and older.

Commit:
Changed org.eclipse.jdt.ui to use any version.

Use the TemplateContext to get the current project.

Commit:
Backwards compatibility for Eclipse <4.10

Commit:
Merge pull request #1 from k-oliver/backwards_compatibility

Backwards compatibility for Eclipse <4.10

Commit:
changed version to 1.1.0.

Automatic escaping of quotes in feature XML.
fixed broken backward compatibility.

Commit:
changed version to 1.1.0

hohwille commented on eclipse-templatevariables-feature/feature.xml:
```diff
@@ -14,68 +14,68 @@
 
    <license>
-                                 Apache License
+      Apache License
```
any reason why this shall not be centered, while the lines below remain centered?
looks odd to me...

k-oliver replied:
Automatically formatted. So no reason.

hohwille commented on eclipse-templatevariables-feature/feature.xml:
```diff
@@ -14,68 +14,68 @@
    1. Definitions.
 
-      "License" shall mean the terms and conditions for use, reproduction,
+      &quot;License&quot; shall mean the terms and conditions for use, reproduction,
```
was that automatically changed by your XML formatter?

hohwille replied:
I do not care much, but just wanted to understand if it is wrong to use regular quotes here (according to my XML understanding this is fine and `&quot;` is only required in double quoted attribute values.)

k-oliver replied:
It was automatically changed by eclipse. Following https://stackoverflow.com/questions/150423/whats-the-accepted-way-of-storing-quoted-data-in-xml BOTH ways are correct. I would suggest to used ``&quot``. Let me know...

hohwille replied:
Sure. Fine this way. Just wanted to understand this...

hohwille commented on eclipse-templatevariables-feature/feature.xml:
```diff
@@ -194,34 +194,39 @@
+         id="net.sf.mmm.eclipse.templatevariables.plugin"
+         download-size="0"
+         install-size="0"
+         version="1.1.0.qualifier"
```
fine. So `0.0.0` was more a hack.
A perfect fix would be to use `${project.version}.qualifier` in combination with maven filtering.

hohwille replied:
You do not need to do this change. I can take it over. You already improved/fixed this versioning and it is fine for the upcoming release `1.1.0` so no need to change for this PR.

hohwille commented on eclipse-templatevariables-plugin/META-INF/MANIFEST.MF:
```diff
@@ -2,14 +2,14 @@ Manifest-Version: 1.0
 Bundle-Name: net.sf.mmm.eclipse.templatevariables.plugin
 Bundle-SymbolicName: net.sf.mmm.eclipse.templatevariables.plugin;singleton:=true
-Bundle-Version: 1.0.0.qualifier
+Bundle-Version: 1.1.0.qualifier
```
same here...

hohwille commented on eclipse-templatevariables-plugin/src/main/java/net/sf/mmm/eclipse/templatevariables/AbstractMavenTemplateVariableResolver.java:
```diff
@@ -66,13 +69,26 @@ private String getMavenProperty(IProject project) {
+            final Object result = method.invoke(context);
+            final IJavaProject javaProject = (IJavaProject)result;
+            return this.getMavenProperty(javaProject.getProject());
+        } catch (final NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
```
Seems as this block above is the important part of the fix. Thanks.

hohwille reviewed (changes requested):
@k-oliver thanks for this PR. I did not change anything here for a long time as I am too busy with other OSS projects...

Some comments on your changes:
* I personally consider it as an anti-pattern to commit eclipse config files into vcs/git
* I was not even aware that I myself committed `.project` and `.classpath` files to git
* seems I was missing a `.gitignore` so thanks for adding one
* IMHO the `.gitignore` should contain `.*` and we should remove all eclipse config files
* FYI: The diffs are larger than necessary as you are using different formatter settings (e.g. indent by 4 instead of 2 spaces). If you are using Eclipse, you might love devon-ide (https://github.com/devonfw/devon-ide/, successor of https://github.com/oasp/oasp4j-ide). The settings can be found here: https://github.com/m-m-m/settings.

Thanks again for your PR. I could also take over and do the suggested changes. Please let me know if you want to update your fork or want to hand-over to me from here.

k-oliver commented:
@hohwille I will update my fork....

Commit:
fixed formatting

Commit:
added .* to gitignore

removed all eclipse settings and project files.

Commit:
revert formatting

k-oliver commented:
@hohwille please review

hohwille commented:
Awesome. Looks great now. @k-oliver 👍 thank you so much.
I will merge it but I also want to test it once. Might take some more days, but nothing more expected to be done from your end. Once merged I will also push our a new release.

hohwille commented:
I cloned your fork repo and switched to your feature branch. When I run the maven build I do get this error:
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:0.22.0:compile (default-compile) on project net.sf.mmm.eclipse.templatevariables.plugin: Compilation failure: Compilation failure:
[ERROR] /projects/mmm/workspaces/main/eclipse-templatevariables-fork/eclipse-templatevariables-plugin/src/main/java/net/sf/mmm/eclipse/templatevariables/AbstractMavenTemplateVariableResolver.java:[79]
[ERROR] 	return this.getMavenProperty(context.getAdapter(IProject.class));
[ERROR] 	                                     ^^^^^^^^^^
[ERROR] The method getAdapter(Class<IProject>) is undefined for the type TemplateContext
[ERROR] 1 problem (1 error)
[ERROR] -> [Help 1]
```

Any ideas? Without this problem being resolved, I can not merge nor release. Did you happen to forget staging some changes in your commit?

k-oliver commented:
I did not try to build with maven. I was simply using the eclipse export which is building against the eclipse running target platform.
I took a look into it and the solution should be updating the eclipse maven repository. I will update my fork...

Commit:
Updated to eclipse 2018-12 maven repository.

Updated to eclipse 2018-12 maven repository and tycho plugin version
to fix the maven build error.

k-oliver commented:
@hohwille I updated my fork. The maven build finished successfully. Please try again.

hohwille reviewed (approved):
@k-oliver Thanks for fixing. The maven build now also works smooth. 👍

hohwille commented:
@k-oliver do you have any hint regarding #2?
I am unable to build since I merged this PR. Maybe trivial to fix for somebody with huge eclipse p2 experience but I am not sure how to proceed. Seems that dependencies from `MANIFEST.MF` such as `org.eclipse.ui` can not be resolved anymore. My first guess is that the updated repository does not contain this anymore. Need to add an additional repo to pom.xml?
